Cleanup in Page::doAfterUpdateRendering() of AppHighlight and sampledTopColor code

smfr opened the pull request:
#### f209e41275a1569d96d2d1638c7ae1f91ef13a78
<pre>
Cleanup in Page::doAfterUpdateRendering() of AppHighlight and sampledTopColor code
<a href="https://bugs.webkit.org/show_bug.cgi?id=283306">https://bugs.webkit.org/show_bug.cgi?id=283306</a>
<a href="https://rdar.apple.com/140138879">rdar://140138879</a>

Reviewed by Aditya Keerthi.

`Page::updateRendering()` and `Page::doAfterUpdateRendering()` need to stay clean, with step-specific
code moved other function, so do this for &quot;sampled page top color&quot; and app highlight code.

Simplify the app highlight &quot;time to restore highlights&quot; code by moving the timing logic into
`AppHighlightStorage`.

* Source/WebCore/Modules/highlight/AppHighlightStorage.cpp:
(WebCore::AppHighlightStorage::shouldRestoreHighlights):
* Source/WebCore/Modules/highlight/AppHighlightStorage.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::restoreUnrestoredAppHighlights):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
(WebCore::Page::computeSampledPageTopColorIfNecessary):
(WebCore::Page::clearSampledPageTopColor):
(WebCore::Page::didChangeMainDocument):
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/286763@main">https://commits.webkit.org/286763@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c815911188ac2aca228bdad29fd2c9cdfd71ecf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77001 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56036 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29916 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/81549 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28280 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79118 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65184 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4332 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60347 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18417 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80068 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50292 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66097 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40658 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47694 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/23594 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/26606 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68813 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23923 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82984 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4381 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2939 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68617 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/4536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66070 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67868 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16928 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11854 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9937 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4327 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7143 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4347 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7782 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6106 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/page/Page.cpp:
```diff
@@ -3262,6 +3245,24 @@ Color Page::sampledPageTopColor() const
+void Page::clearSampledPageTopColor()
+{
+    if (m_sampledPageTopColor) {
+        m_sampledPageTopColor = std::nullopt;
```
Could use `std::exchange` here.

pxlcoder replied:
(but I understand we're just moving existing code, so also fine to keep as-is)

smfr committed:
Cleanup in Page::doAfterUpdateRendering() of AppHighlight and sampledTopColor code
https://bugs.webkit.org/show_bug.cgi?id=283306
rdar://140138879

Reviewed by Aditya Keerthi.

`Page::updateRendering()` and `Page::doAfterUpdateRendering()` need to stay clean, with step-specific
code moved other function, so do this for "sampled page top color" and app highlight code.

Simplify the app highlight "time to restore highlights" code by moving the timing logic into
`AppHighlightStorage`.

* Source/WebCore/Modules/highlight/AppHighlightStorage.cpp:
(WebCore::AppHighlightStorage::shouldRestoreHighlights):
* Source/WebCore/Modules/highlight/AppHighlightStorage.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::restoreUnrestoredAppHighlights):
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::doAfterUpdateRendering):
(WebCore::Page::computeSampledPageTopColorIfNecessary):
(WebCore::Page::clearSampledPageTopColor):
(WebCore::Page::didChangeMainDocument):
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/286763@main
